File: fsrs-java/src/main/java/dev/bee/fsrs/FsrsMemoryState.java
package dev.bee.fsrs;

/**
 * Immutable FSRS memory state for a reviewed item.
 */
public record FsrsMemoryState(double stability, double difficulty) {
    public FsrsMemoryState {
        if (!Double.isFinite(stability) || stability <= 0.0) {
            throw new IllegalArgumentException("stability must be finite and positive");
        }
        if (!Double.isFinite(difficulty) || difficulty < Fsrs.MIN_DIFFICULTY || difficulty > Fsrs.MAX_DIFFICULTY) {
            throw new IllegalArgumentException("difficulty must be finite and in [1, 10]");
        }
    }

    @Override
    public String toString() {
        return "FsrsMemoryState{stability=" + stability + ", difficulty=" + difficulty + '}';
    }
}

File: fsrs-java/src/main/kotlin/dev/bee/fsrs/FsrsMemoryState.kt
package dev.bee.fsrs

/**
 * Immutable FSRS memory state for a reviewed item.
 */
@JvmRecord
data class FsrsMemoryState(
    val stability: Double,
    val difficulty: Double,
) {
    init {
        if (!stability.isFinite() || stability <= 0.0) {
            throw IllegalArgumentException("stability must be finite and positive")
        }
        if (!difficulty.isFinite() || difficulty < Fsrs.MIN_DIFFICULTY || difficulty > Fsrs.MAX_DIFFICULTY) {
            throw IllegalArgumentException("difficulty must be finite and in [1, 10]")
        }
    }

    override fun toString(): String =
        "FsrsMemoryState{stability=$stability, difficulty=$difficulty}"
}
